operators/pm.go

```diff
@@ -17,38 +17,47 @@ package operators
 import (
 	"strings"
 
-	"github.com/cloudflare/ahocorasick"
 	"github.com/corazawaf/coraza/v2"
+	ahocorasick "github.com/petar-dambovaliev/aho-corasick"
 )
 
 // TODO according to coraza researchs, re2 matching is faster than ahocorasick
 // maybe we should switch in the future
 // pm is always lowercase
 type pm struct {
-	matcher *ahocorasick.Matcher
-	// dict is used for capturing
-	dict []string
+	matcher ahocorasick.AhoCorasick
 }
 
 func (o *pm) Init(data string) error {
 	data = strings.ToLower(data)
-	o.dict = strings.Split(data, " ")
-	o.matcher = ahocorasick.NewStringMatcher(o.dict)
+	dict := strings.Split(data, " ")
+	builder := ahocorasick.NewAhoCorasickBuilder(ahocorasick.Opts{
+		AsciiCaseInsensitive: true,
+		MatchOnlyWholeWords:  false,
+		MatchKind:            ahocorasick.LeftMostLongestMatch,
+		DFA:                  true,
+	})
+
 	// TODO this operator is supposed to support snort data syntax: "@pm A|42|C|44|F"
-	// TODO modsecurity uses mutex to queue ahocorasick, maybe its for a reason...
+	o.matcher = builder.Build(dict)
 	return nil
 }
 
 func (o *pm) Evaluate(tx *coraza.Transaction, value string) bool {
-	value = strings.ToLower(value)
-	matches := o.matcher.MatchThreadSafe([]byte(value))
-	for i := 0; i < len(matches); i++ {
-		if i == 10 {
-			return true
+	if tx.Capture {
+		matches := o.matcher.FindAll(value)
+		for i, match := range matches {
+			if i == 10 {
```

Comment:
I know it isn't part of the PR but it would be really cool to get a hint on what this 10 means.

Reply:
Excellent question :) This is another constant that we should be naming properly.

10 is the maximum captures you can have (originally in modsecurity). This comes from https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#capture
```
Up to 10 captures will be copied on a successful pattern match, each with a name consisting of a digit from 0 to 9. The TX.0 variable always contains the entire area that the regular expression matched. All the other variables contain the captured values, in the order in which the capturing parentheses appear in the regular expression. 
```